### ISSUE TITLE
wxSocketBase::Peek() works like Read(), bytes become unavailable

### ISSUE THREAD

**Issue by GeorgyStolbovoy** [wxWidgets/wxWidgets#23594]
### Description

I am trying to communicate with a server application over UDP protocol. This application responds sequentially in several chunks. Each chunk has some important data and much more. These important data must be extracted before reading the whole chunk. For this purpose I used wxSocket::Peek(). As it turned out, the subsequent call to wxSocket::Read() reads the next chunk! The chunk that Peek() was looking at just got lost! Parts of my code:
```c++
// Initialization in constructor
	Bind(wxEVT_SOCKET, &Recieve, this, socket_id);
	wxIPV4address addr;
	addr.Hostname(wxT("localhost"));
	addr.Service(16801);
	socket = new wxDatagramSocket(addr);
	socket->SetEventHandler(*this, socket_id);
	socket->SetNotify(wxSOCKET_INPUT_FLAG);
	socket->Notify(true);
// ...
// Responce handler
void Frame::Recieve(wxSocketEvent& event)
{
	std::vector<uint8_t> packet;
	uint8_t peek[7];
	socket->Peek(&peek, 7); // Here
	uint16_t w = *reinterpret_cast<uint16_t*>(&peek[3]); // width and height
	uint16_t h = *reinterpret_cast<uint16_t*>(&peek[5]); // of a image
	packet.resize(7 + w*h*2);
	for (uint8_t i(0); i < peek[1]; ++i) // peek[1] is the number of chunks
	{
		socket->Read(packet.data(), packet.size()); // OOPS, this is next chunk
// ...
``` 
If Peek() is not used, then Read() will read the first chunk. If I call Peek() twice, then Read() will read the third chunk. Thus, Peek() changes the state, although it should not do this!

### Platform and version information

- wxWidgets version:  3.2.2.1
- wxWidgets port: wxMSW
- OS: Windows 10

**Comment by BrianNixon** [comment]
On the one hand: this is how the datagram protocol is meant to work. When a single `recvfrom()` does not consume an entire datagram, the remainder is lost.

On the other hand: wx tries to present datagram sockets as though they are streams. To avoid this data loss, `Peek()` on a datagram socket needs to use a buffer big enough to capture the largest possible datagram, not just the number of bytes the caller requested.

**Comment by BrianNixon** [comment]
Here’s a repro. Note that the `LastCount()` checks will fail without #23598; the problem being reported here is the failing `recvbuf[1] == sendbuf1[1]`.

```cpp
TEST_CASE("wxDatagramSocket::PeekPartialDatagram", "[socket][dgram]")
{
    wxIPV4address addr;
    addr.LocalHost();
    addr.Service(27384);// Arbitrary port number
    wxDatagramSocket sock(addr);

    // Send 2 datagrams
    unsigned int sendbuf1[2] = {1, 2};
    sock.SendTo(addr, sendbuf1, sizeof(sendbuf1));
    unsigned int sendbuf2[2] = {3, 4};
    sock.SendTo(addr, sendbuf2, sizeof(sendbuf2));

    // Peek the first word
    unsigned int peekbuf[1] = {0};
    sock.Peek(peekbuf, sizeof(peekbuf));
    CHECK(sock.LastCount() >= sizeof(peekbuf));
    CHECK(peekbuf[0] == sendbuf1[0]);

    // Read the whole of the first datagram
    unsigned int recvbuf[2] = {0};
    sock.Read(recvbuf, sizeof(recvbuf));
    CHECK(sock.LastCount() >= sizeof(recvbuf));
    CHECK(recvbuf[0] == sendbuf1[0]);
    CHECK(recvbuf[1] == sendbuf1[1]);
}
```

**Comment by BrianNixon** [comment]
Candidate fix in https://github.com/BrianNixon/wxWidgets/commit/54f5701f0a97e0535890f4a1ae7891549bbe7b97.

(Can I make a PR out of that, while preserving the dependency on #23598, which hasn’t been merged? I don’t understand Git…)

**Comment by vadz** [comment]
Thanks for the fix (and I've actually failed to notice that it wasn't included in #23598 yet, when I had already written this there)! I think you should simply add this commit to the same PR, i.e. just include it in your `dgramsocket_short_read` branch and push it to GitHub -- PR will update when any new commits are pushed to the branch it is based on.

TIA!

**Comment by BrianNixon** [comment]
#23598 is now closed in favour of #23604, so I’ve added the fix+​test there. I’d just kept them separate as they’re not directly related; plus I’m not convinced this is a great fix
